### ISSUE TITLE
CSS: Consider using a smart server side CSS parser

### ISSUE THREAD

**Issue by povrazor** [ludumdare/ludumdare#87]
Up until now we've been using PHP files that generate CSS. This is okay, but the CSS syntax is still junk.

Some parsers out there offer a much nicer CSS syntax. There are limitations to consider. PHP does give us a lot of added flexibility, but practically we haven't been using it. The majority of work in `.css.php` files has been color mixing. A real tool may do a better job overall.

Introducing a CSS parser has the added benefit of not having to change the file extension, meaning CloudFlare will be able to figure out it should be cached.

The downside is that it has to be implemented both on the Apache (toolchain) and LSWS (server) side. Ultimately it's probably worth it though.

**Comment by BillyWM** [comment]
The two main contenders are Less and Sass. These days it seems Sass is the more popular one. Either would be an advantage over doing it manually in PHP. Both have niceties like variables, nesting rules, an 'include' system, functions, etc (Functions and logic are where Sass wins, IMO. They're more powerful and yet easier to write). Color calculation is something they're commonly used for.

For caching purposes, typically you'd include .sass -> .css conversion and minification as a step in your build process and then serve the minified css.

**Comment by BillyWM** [comment]
All of your constants can become Sass variables. e.g.

    const MOBILE_QUERY = "(max-width : 600px)";

becomes

    $MOBILE_QUERY: "(max-width : 600px)";

and can be substituted in place the same as now, without the line noise of jumping in and out of <?= php mode ?>

The lookup table in theme-hub-.css, however, looks like it can be calculated. Is this calculated somewhere already? It could be done in-place, instead.

    background:<?=PAL_RED[COL_MUTE_LIGHT]?>;

becomes

    $base-color: #622;
    background: mute-light($base-color)

where mute-light is a custom function you've written.

**Comment by povrazor** [comment]
I have been looking in to Sass, but in an unusual way. What I wanted to do was configure the server(s) to automatically run `.css` files through the SASS parser, as you typically do with `.php` and other server side scripts.

* Development side: It would mean you could test changes natively, without the need to recompile.
* Deployment side: Because I run our sites behind a reverse proxy (CloudFlare), they would cache a the output `.css` file.

To my surprise there was no FastCGI SASS interpreter. It might be because Sass is a Ruby tool, but there is an official C library that does Sass parsing. So a few weeks ago I started experimenting with this, and I pretty-much know exactly what to do to get this working, I just occupied by other things and wasn't able to finish.

https://github.com/povrazor/sass-cgi

The wall I was unfortunately rather fundamental (i.e. getting the FastCGI script to execute). I think I have my test server configured correctly. It hangs exactly where the script should be running. But again, I haven't had a chance to pick it up and investigate since there.

**Comment by povrazor** [comment]
After a bit of homework, I'm leaning away from Sass and leaning more towards using tools in the PostCSS suite.

https://github.com/postcss/postcss

That said, I'm focused on backend at the moment, so it's not something I'm too concerned with right now.

**Comment by djdduty** [comment]
Are css files changed dynamically during the application serving period?. 

Rather than add something new to the toolchain that would build and serve some css template files, just add a build step to your development. During development you can run an application created in any language you want that will listen for file changes and immediately re-build Sass / output regular css files.

For my use, I use a node package called webpack. During development webpack runs and listen for file changes on js or sass(scss in my case) files, upon any change it re-builds them into packaged distribution files that my web application is already pointing at, expecting to serve them like regular ol' js or css files (they are at this point). If you use a separate cdn, you can just upload those dist files to the cdn, they are regular css files.

During deployment just run your packer / Sass builder once to create dist files or include the built / minimized output files into the repository to avoid the deployment step.

**Comment by povrazor** [comment]
> Are css files changed dynamically during the application serving period?

No. Or rather, they wont be. 

Currently the CSS files are PHP files, which does give them the luxury of variables. For testing, parameters can be tweaked as part of the request, but in practice it's clumsy. Way back, I had a lot of requests for being able to customize the theme, but in practice this doesn't make CSS any nicer to work with.

I'm more interested in writing shorter, concise, localized (not translation) CSS code. PHP gave us variables, but things like nesting/grouping would also be handy, once a style is finalized.

AutoPrefixer is a great example of what I see as genuinely useful, and it comes as part of the PostCSS suite. Hence why I'm more leaning that direction.

https://github.com/postcss/autoprefixer

Anyway, like I mentioned on Twitter I'm not currently putting too much focus on frontend (I'm all backend and server side at the moment). I am casually investigating things as a change of pace, but like I said, it's casual at the moment. I didn't like how development worked with the WIP prototype from last summer. I know it can be better.

For a very rough look at what I'm thinking, on the bottom of the style guide page is a random scattering of notes.

https://github.com/ludumdare/ludumdare/wiki/Style-Guide

What I don't say there is that the frontend needs a toolchain, and I haven't decided what I like there yet. Tools like PostCSS and AutoPrefixer definitely seem like the right direction when it comes to CSS (probably some GULP build file), but for pages and layout I'm not entirely decided there. The fewer distinct languages and dialects, the better.

This is getting off topic, but when I say backend, I'm referring to the REST API and all the database access goodness, local caching, and optimization there. A large part of the site will be moved entirely to restful calls, but for search engine indexing, your regular URLs still need to still spit out something. 

And as the code base has grown, I'm not entirely happy with how some parts are done (for example, 30+ database table create queries in one script file, and a 2nd file that needs to be updated with deletes). A topic I hope to cover in the style guide is locality, the idea that things that relate to a topic should be in one place. But there's a give and take there. It's easier to scroll down/search in a single file then it is to search across multiple files. In practice a single file for everything isn't practical, but a single folder could work (i.e. everything related to users in `/modules/user`, the DB create script, destroy script, the backend query calls, the JS code for calling the API, etc). 

I am happy with the elements (PHP 7, APCu, JavaScript, CSS, MariaDB, Vagrant, OpenLiteSpeed, Ubuntu), it's just tying them together the right way.

**Comment by djdduty** [comment]
On the front end topic, as is pertaining to this topic, a gulp build file is exactly what I'm talking about.

You can use node-sass with postcss together to create a Sass compiler/autoprefixing build step, and you can even auto-listen to file changes during development to rebuild sass to css output files (autoprefixing doesn't matter so much in a dev environment). Create a gulp file that you will run as a deployment build step to output minified, prefixed css dist files compiled from sass and BOOM, done! That front really isn't too complicated.

As far as database scripts, have you looked into any sort of migration system? It would be immensely helpful if you're sticking with a relational db.

**Comment by djdduty** [comment]
The only thing you would have to add to your toolchain, and for development only, is node.
With node you can then add node-sass and postcss to your package dependencies.

### During development ###
If you want to watch the sass files for changes and rebuild the css automatically
`node-sass -w static/sass/ -o static/build/css/`

### During deployment ###
`node-sass static/sass/ -o static/partial/css/`
`postcss --use autoprefixer static/partial/css/*.css -d static/build/css/`
Obviously this step would be automated into one command via gulp and wouldn't create the intermediary partial folder, I just don't know the exact usage of gulp to say here.

Now your static/build/css has regular ol' css files that you can use, upload to cdn, whatever.

**Comment by povrazor** [comment]
I'm not entirely convinced of Sass yet. Several of its key features are becoming part of the core CSS 4 standard, and I'm more likely to prefer something that's future proof. WordPress bit us in the arse many times over the past decade, and things that were hot a decade ago like Flash are becoming rarer today. 

CSS is here to stay (it's the web). JavaScript is here to stay (it's also the web). Ruby, and a custom parser written in Ruby (Sass), it's hard to say. It may become the Python of the next decade (i.e. around but not really changing).

But I'll have to see. I haven't looked as closely at the CSS 4 proposals yet, what Microsoft/Google/Mozilla/Apple think yet (whose in), and what features of parses like Sass and Less are genuine improvements to CSS development.

And yes there will be some data migration, but that's a whole other topic.

**Comment by djdduty** [comment]
In my opinion, it's good to rely on browsers implementing brand new functionality as little as possible. It's probably not a great idea to put all your chips into css4 since that LARGELY depends on the browser that is rendering your page. With something like Ludum Dare you have a lot of users with a whole ecosystem of different browsers and different versions and you don't want to be poly filling missing features, as it is you'll have to poly fill flex-box anyway for certain browsers. ( please tell me you're using flex box :D ). 

Sass gives you css that works now, that will work in the future, and it's definitely not going any where. Sass also definitely is NOT a ruby only thing it's used everywhere now in many different environments and contexts that just happened to be where it started. My company uses it in our python project via the webassets extension, we pack it up using node-sass for our php project. It can't go anywhere there's too many people using it and relying on it.

As I said though, you can 100% use postcss WITH sass, Sass is basically just css with nesting, inheritence, importing, and variable support, you will be able to write css4 inside of Sass files if you really want. There's really no downside to using it except the build step, but you're adding the build step anyway with postcss so might as well take one extra step while you're at it.

**Comment by povrazor** [comment]
> In my opinion, it's good to rely on browsers implementing brand new functionality as little as possible.

Oh sure. I meant syntactically. Sass, Less, and CSS 4 are just lists of features. No matter which way you go, you're still beholden to the limitations of CSS 2 and 3. I may still favour Sass in the end, but I want the preprocessor to favour the coding style. I'm using CSS 4 as an example, but PostCSS has hundreds of plugins, and it itself is designed to make writing new parsers easy. What I'm trying to say is I'm not ruling out Sass alternatives, at least not yet.

For example, of the style guides out there, I'm currently favouring one called RSCSS.

http://rscss.io/

There's a brevity and descriptiveness to it I like versus BEM and ECSS, but I haven't had a chance to see what I think of it in practice. 

One coding style document I was reading made reference to bugs in Sass regarding media queries. I can't remember where I saw it, but given how splintered the Sass spec is (original Ruby Sass vs. everything based on libsass), it should be tested.

I'm just saying this a fairly important decision with a bunch of repercussions. ;)

**Comment by amcgregor** [comment]
My… writing style can often come across as overly hostile, so I will apologize up-front.  I'm sorry if my enthusiasm for web development forces me into any hasty wording, or over-triteness.  I post this comment with the best of intentions, to help inform and educate, not to merely confront.  (Nor am I trying to sell you on Python, but this is where my main experience lies.  Just want to show what web dev can look like.)

TL;DR — Please, please, please use [SCSS](http://sass-lang.com/guide), either [autoprefixer](https://github.com/postcss/autoprefixer#readme) ([recommended by Google](https://developers.google.com/web/tools/setup/setup-buildtools#dont-trip-up-with-vendor-prefixes), used by Twitter) or [postcss](https://github.com/postcss/postcss#readme) which includes it, and a build automation tool of some kind, with either a build step for deployment or git commit hook to always check-in the latest compiled/minified versions during development, and built-in file watching / live reloading capabilities while developing.

For a concrete example, your Style Guide wiki page RSCSS example: (split on two lines to illustrate)

```css
.search-form .-variant {}
.search-form > .button {} 
```

vs.

```scss
.search-form {
    &-variant {}
    
    > {
        button {}
    }
}
```

Removal of repeated selector prefixes.  Which is more readable and structured?

---

> A real tool may do a better job overall.

I guarantee you it will.  PHP is an HTML template engine, not a CSS generation engine.  It has no optimizations, shortcuts, helpers, or integration to assist in the task of building CSS.

> Introducing a CSS parser has the added benefit of not having to change the file extension, meaning CloudFlare will be able to figure out it should be cached.

This is irrelevant when using preprocessors correctly.  There are several options: run a watcher during development to automatically recompile/minify, add it as a git commit hook, or have a build stage prior to deployment.  During development, SASS can be live-reloaded in your browser while you work using the first approach, massively decreasing turnaround time on making design changes.  `.map` file generation means you can even _use_ your browser's developer panel to edit your SASS, with changes being quite live.

> I have been looking in to Sass, but in an unusual way. What I wanted to do was configure the server(s) to automatically run .css files through the SASS parser, as you typically do with .php and other server side scripts.

This is an unfortunate approach to investigating the tool; using it in a non-standard, non-recommended way is a sure fire recipe to get the wrong impression, get frustrated, and give up on the tool.  It is not meant to be used that way at all.  (Sure, you can, but you're making life harder on yourself.)  Notably, because the potential build process for your CSS might include automatic vendor prefix addition, automatic sprite sheet construction, automatic data-uri construction, combination, and minification, amongst others.  You want the final asset delivered to the client to be a single static `.css` file, generally, not a hodgepodge of generated-on-demand (slow) needlessly dynamic individual files.

> To my surprise there was no FastCGI SASS interpreter. It might be because Sass is a Ruby tool, but there is an official C library that does Sass parsing.

SASS is a universal tool.  (I use `python-libsass` which executes the C library routines natively, using the Python asset build automation tool WebAssets to manage such generation, minification, etc.  Even contributed back the ability for `@import` directives to be serviced by Python callbacks…)  There is no FastCGI SASS interpreter because it is an explicitly bad approach to using the tool, losing out on a large majority of the benefits of the tool.

> After a bit of homework, I'm leaning away from Sass and leaning more towards using tools in the PostCSS suite.

They are not mutually exclusive, and work exceptionally well _together_.  One does not replace or obviate the other, unlike postcss including autoprefixer.

> Currently the CSS files are PHP files, which does give them the luxury of variables.

So does BASH shell script.  ;)  This still doesn't make it a good approach when richer, more task-specific tools exist that are better than using a general purpose language, or simple template engine, in every measurable way.

> The only thing you would have to add to your toolchain, and for development only, is node.

Actually… you don't even need that, if you use a language like Python which supports running JavaScript directly within your application's process.  My own codebases run JavaScript as Python (letting me use Node tools like autoprefixer natively via an embedded `ducktape` interpreter, possibly one is available for PHP?), and let me write my front-end JavaScript _as_ Python using transpilers (transcrypt or others).  Turtles all the way down, one language syntax to keep in your head-space, one set of static analysis tools, etc.

> CSS is here to stay (it's the web). JavaScript is here to stay (it's also the web). Ruby, and a custom parser written in Ruby (Sass), it's hard to say. It may become the Python of the next decade (i.e. around but not really changing).

Fearing that a tool used by millions worldwide will go away spontaneously is not a reasonable fear.  WordPress is hardly comparable, as it's not a build tool, it's an entire application suite.  Even when SASS added SCSS syntax, they didn't break older use, it's open source so you can always build your own copy, etc.  Your comment on Python demonstrates a substantial lack of knowledge about Python.  (3.5 adding end-to-end async support and async flow control such as `async def` or `async for` being only one example of recent change.  The times, they are a-changing! ;)

> I haven't looked as closely at the CSS 4 proposals yet, what Microsoft/Google/Mozilla/Apple think yet (whose in), and what features of parses like Sass and Less are genuine improvements to CSS development.

What CSS 4 "in the future" will provide is irrelevant in the face of a tool (SASS) that exists in stable production-ready form today, and CSS being text, and SASS a text processor for CSS, it'll be compatible anyway.  Nested rulesets.  Variables.  Functions.  Mix-ins / templates.  Ease of component-driven-design.  Inclusions.  The ability to define callbacks and SASS functions in the general purpose language of your choice and export them to your SASS.  This is just a small list of improvements over bare CSS.

> Sass, Less, and CSS 4 are just lists of features. No matter which way you go, you're still beholden to the limitations of CSS 2 and 3.

You seem to misunderstand the use of a number of the components listed so far.  The things you have reimplemented (`.css.php` gives me the willies ;) are solved problems, better solved using one of the existing packages.  There are several reasons why this is the case: "feature lists" is a casual dismissal, here, and "feature lists" aren't the point.  Increasing developer productivity is.  Even more useful is the increased headcount of potential contributors who know how to work with these existing solutions.  I recommend SCSS (a slightly more CSS-flavoured format for SASS), as does most of the industry.  This isn't because SASS adds any end-user features beyond CSS ("beholden to the limitations" is… an interesting phrasing!  I approve. ;) but instead because it makes construction of your CSS a simplified, boilerplate-reduced, duplicated-selector-eliminated, no-more-by-hand-processing (ref: sprites, data-uri, functions for colours and such) joy, with high likelihood that random off-the-street front end developers are already comfortable working with it.

Similarly, use of `autoprefixer` (a standard part of `postcss` but usable independently) completely eliminates the need to write vendor-specific prefixes in order to support these standard CSS features, further reducing the developer headspace needed to safely and correctly make changes or additions to your site design.

Thank you for taking the time to read my feedback on this situation.

**Comment by povrazor** [comment]
**First** I should mention this is a necro-thread, originally posted over a month ago, before I started looking in to this. A TODO item. I am not focused on front end, so I've only been casually looking at this stuff.

**Second**, my fault since I didn't make it clear, I canned the SassCGI plugin after learning about Compass and other file scanners, but more importantly after learning about PostCSS. There is *A LOT* under the PostCSS banner, a lot of it is CSS 4, and what I've done a poor job saying is I want to be sure I know what PostCSS has to offer before I finalize anything. I very briefly looked at CSS 4, and I saw the things I was looking for that my lame PHP setup lacked. I don't come from a web-dev background, so I have no history with the big 3 (Sass, Less, PostCSS), no prejudice with them, so I'd like to be thorough. 

This article by the author of a Sass book saying "Oh, it turns out you don't need Sass" got me started down this rabbit hole:

https://benfrain.com/breaking-up-with-sass-postcss/

**Third**, I have been looking at Less and Sass, but admittedly I am leaning toward Less. I spent the past few hours doing a raw feature/syntax compare, since they do the exact same thing. I've discovered there is a lot of misinformation regarding Less. Here's one such article trying to clear it up:

https://getcrunch.co/2015/10/08/less-the-worlds-most-misunderstood-css-pre-processor/

I do like how variable evaluation works in Less versus Sass (i.e. lazy). It's more flexible. The dialect is starting to grow on me as well. 

I am not done my homework though.

> Your comment on Python demonstrates a substantial lack of knowledge about Python

Thanks.

A decade ago, Python and PyGame were a common and popular way to make Ludum Dare games. Nearly a quarter, if not more used it. Today, it's less than 1%. 

Also, software that added integrated scripting support around Y2K tend to use Python (Blender). Software that added integrated scripting closer to 2010 and beyond tend to use Lua.

That's what I'm talking about when I talk about the fall of Python. It's not a _hot_ language anymore.

**Comment by BillyWM** [comment]
Ludum Dare goers probably moved away from PyGame because it's just not a very good engine.

Flixel + FlashPunk were the LD favorite for a while, then Unity now right? The Unity tooling just blows everything else out of the water. Doesn't really mean its supported languages are the best.

As for scripting integration, Lua is _designed_ for integration. You compile it right in and expose it through a C api.

-------------------------

Getting back to CSS....

Here's the best resource for looking up CSS compatibility:

http://caniuse.com/#tables
e.g http://caniuse.com/#feat=css-animation

**Auto-prefixing**: I don't see this as very important actually, but you have multiple tools that can provide it. You don't have to buy in to one environment. I do a lot of frontend work in my job and I just take a quick look at caniuse when I'm unsure of needed prefixes.

**CSS4**: Technically there's no "CSS4". CSS is now split into modules and there's a "selectors level 4" module. But in any case we're overthinking things here. Your existing CSS is pretty simple, even with variable substitutions. 

On the topic of variables, look how ugly native CSS variables are:

https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_variables

Not supported on IE, Android below 5, older or not-updated iPhones....
http://caniuse.com/#feat=css-variables

So pre-processors provided this feature for years before a spec arrived and they'll still be the best choice for years to come. I wouldn't bet everything on any future spec.

**Comment by BillyWM** [comment]
Since you don't have a web dev background, I'd say don't overthink it and just try one out. Don't worry about picking the "perfect" preprocessor, especially for such relatively simple CSS as you have now.

**Comment by amcgregor** [comment]
Heh, seemed pretty active for a necrothread, but apologies for any incidental summoning from the dead.  ^_^;

> @povrazor: A decade ago, Python and PyGame were a common and popular way to make Ludum Dare games. Nearly a quarter, if not more used it. Today, it's less than 1%.

So your experience with Python is a decade old, with a horrendously engineered tool.  I don't see how any of this is relevant to web development and not just more distraction from just picking something and getting moving.  My original statement stands.

> @BillyWM: Auto-prefixing: I don't see this as very important actually…

You mention caniuse.com, which means _manual labor_ and headspace, which is silly!  autoprefixer (again, independently or as part of postcss, I don't really care which) uses the caniuse.com data itself to _automate_ that labour.  I honestly can not comprehend the friction, here.

> @povrazor: I have been looking at Less and Sass, but admittedly I am leaning toward Less.

SASS benefits from a substantially nicer (less hacky to get "real work done") syntax for functions and flow control (conditional statements).  When I used LESS, my function declarations involving conditionals looked like Christmas trees.  Here's a counter-article: https://css-tricks.com/sass-vs-less/  (But we can link counter-counter and counter-counter-counter articles to eachother until the end of days, that won't provide any progress on actually getting things done.)
